tools/ci/commitGeneratedDocs.ps1

```diff
@@ -0,0 +1,138 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Commit a generated documentation to the gh-pages documentation branch
+# For the master branch, this commits the documentation at the root of
+# the branch. For other branches, this commits it under the versions/ folder.
+
+param (
+    [Parameter(Position=0)]
+    [string]$SourceBranch
+)
+
+# Silence errors to avoid PS failing when git commands write to stderr.
+# See https://github.com/microsoft/azure-pipelines-yaml/issues/306
+function RunWithoutFailingOnStdErr($Command, $DisplayName = $Command) {
+    $oldEAPreference = $ErrorActionPreference
+    $ErrorActionPreference = "SilentlyContinue"
+
+    $out = Invoke-Expression $Command -ErrorVariable err
+
+    if ($LastExitCode -ne 0)
+    {
+        $ErrorActionPreference = $oldEAPreference
+        $msg = "Command '$DisplayName' returned an error"
+        Write-Host $msg
+        Write-Host $out
+        Write-Host $err
+        Write-Host "##vso[task.complete result=Failed;]$msg"
+        exit 1
+    }
+    $ErrorActionPreference = $oldEAPreference
+    return $out
+}
+
+# Strip the source branch from refs/heads/ if any.
+# In general this is coming from Build.SourceBranch which has it.
+# Build.SourceBranchName cannot be used because it strips also the subfolders.
+# https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables
+$SourceBranch = ($SourceBranch -Replace "^refs/heads/","")
+Write-Host "Source branch: '$SourceBranch'"
+
+# Create some authentication tokens to be able to connect to Azure DevOps to get changes and to GitHub to push changes
+Write-Host "Create auth tokens to connect to GitHub and Azure DevOps"
+$Authorization = "Basic " + [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes("${env:GITHUB_USER}:${env:GITHUB_PAT}"))
+
+# Check that source branch exists
+# Note that the Azure DevOps checkout is a specific commit not a branch,
+# so the local repository is in detached HEAD. To test if the branch exists,
+# use the remote one from the 'origin' remote.
+Write-Host "Resolve source branch commit SHA1"
+$output = ""
+Invoke-Expression "git rev-parse --verify `"refs/remotes/origin/$SourceBranch^{commit}`"" | Tee-Object -Variable output | Out-Null
+if (-not $output) {
+    Write-Host "Unknown branch '$SourceBranch'"
+    Write-Host "##vso[task.complete result=Failed;]Unknown branch $SourceBranch."
+    exit 1
+}
+
+# Get info about last change associated with the new generated docs
+Write-Host "Get the SHA1 and title of the most recent commit"
+$commitSha = git log -1 --pretty=%H
+$commitTitle = git log -1 --pretty=%s
+Write-Host "${commitSha}: $commitTitle"
+
+# Clean the _docs/ folder to avoid any interaction with a previous build if the agent is not clean
+Write-Host "Clean output folder '_docs/' if it exists"
```

Comment:
Definitely looks like a candidate to extract as a function. "Delete all children of X"